### PR TITLE
Fix #15090 [1.4] Added dbt existing glossary and tier ingestion

### LINKED ISSUE #15090: dbt Enhancements
# [1.3.2] - JSON Schema & Parsing Improvements
- [x] Add `type` fields in the json schema of each individual config of dbt (local, http, s3, gcs, azure, cloud)
- [x] Handle the appropriate error in case of type mismatches

# [1.4] - Tags & Glossaries
- https://github.com/open-metadata/OpenMetadata/issues/9031
- https://github.com/open-metadata/OpenMetadata/issues/13293

We currently sync dbt tags into OM by creating a new classification `DBTTags` and adding all that info as tags inside.
What we need to figure out here is a way to directly link `dbt` tags into **existing** tags/tiers/glossaries in OpenMetadata. Example:

```yaml
tables:
- name: DATA_TABLE
  description: Data ,
  columns:
  - name: gross_revenue
    description: column description
    meta:
      openmetadata:
      # DO NOT create anything new in OM, just link to existing items
        - type: GlossaryTerm
          name: BusinessGlossary.GrossRevenue
        - type: Classification
          name: Tier.Tier1
```

# [1.4.1] - dbt run details
- https://github.com/open-metadata/OpenMetadata/issues/14065

We need to figure out how to link datamodels with relevant information like:
- past executions & status
- last time it was refreshed

We have 2 different topics here:
- How to integrate dbt cloud metadata (schedule, runs, jobs, etc.) -> This can become a new Pipeline Service
- How to figure out - when using dbt core - when each model was refreshed etc. (https://docs.getdbt.com/reference/artifacts/run-results-json) -> These are extra properties to add to a `DataModel` Entity, be it a `dbt` model, or `DashboardDataModel`: add `lastRefreshed`, `executions` (similar to `Pipeline` Entity executions)

# [1.5] - Semantic Layer
- https://github.com/open-metadata/OpenMetadata/issues/12911

How to integrate GENERALLY "Semantic Layer" data, be it from dbt metrics/exposures, Tableau Metrics, etc.
- Would it make sense to add a "Metric" field to `GlossaryTerm` Entity that can be an SQL expression, Python code that computes the metrics, etc.

# [1.5] - dbt Hooks
- Can we build a GitHub action that validates changes in dbt vs. metadata in OM?

# [1.3.2] - Documentation
- https://github.com/open-metadata/OpenMetadata/issues/7113
- [x] remove https://docs.open-metadata.org/v1.3.x/sdk/python/ingestion/dbt
- [x] Follow the approach as with any connector (UI vs. run externally with new step-wise components)

# Backlog
- [ ] **P2** -  For dbt cloud, the metadata required can be gathered using a metadata API from dbt cloud. Check if it is feasible to implement it instead of current approach of getting dbt artifacts.
- [ ] **P2** https://github.com/open-metadata/OpenMetadata/issues/12910
- [ ] **P2** https://github.com/open-metadata/OpenMetadata/issues/12095

### PR DESCRIPTION
**Pull request by SumanMaharana**
### Describe your changes:

Fixes #15090 

#
### Type of change:

- [ ] Bug fix
- [x] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [ ] I have commented on my code, particularly in hard-to-understand areas. 
- [ ] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

### CODE REVIEW

File: ingestion/src/metadata/ingestion/source/database/dbt/metadata.py [hunk 1]
```diff
@@ -398,6 +407,33 @@ def yield_data_models(
                             include_tags=self.source_config.includeTags,
                         )
 
+                    dbt_table_glossaries_list = None
+                    dbt_table_tier = None
+                    if manifest_node.meta and manifest_node.meta.get("openmetadata"):
+                        if manifest_node.meta["openmetadata"].get("glossary"):
```
[the remark this answers is not in the document]
Comment (SumanMaharana) [reply ingestion/src/metadata/ingestion/source/database/dbt/metadata.py]: Added a DbtMeta pydantic model for glossary and tiers

File: ingestion/src/metadata/ingestion/models/patch_tags.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from pydantic import BaseModel
+
+from metadata.generated.schema.entity.data.table import Table
+from metadata.generated.schema.type.tagLabel import TagLabel
+from metadata.ingestion.api.models import Entity
+from metadata.ingestion.models.table_metadata import ColumnTag
+
+
+class OMetaGlossariesAndTiersData(BaseModel):
+    entity: Optional[Type[Entity]]
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/models/patch_tags.py:25]: hi sorry why is this needed? we were already adding tags from dbt right

**Review by pmbrull — COMMENTED** [review]
left a question there. We should also review which tests we have for dbt and validate we are properly parsing manifests and getting the required info

File: ingestion/src/metadata/ingestion/models/patch_tags.py (hunk 2 above)
Comment (SumanMaharana) [reply ingestion/src/metadata/ingestion/models/patch_tags.py:25]: we were adding or creating tags from dbt as the entity type `Tag` specifically.
but in case of adding glossaries it's entity type was `GlossaryTerm` so created this for patching glossaries.

File: ingestion/src/metadata/ingestion/source/database/dbt/metadata.py [hunk 3]
```diff
@@ -578,6 +617,44 @@ def parse_data_model_columns(
...
+                            )
+                except Exception as exc:  # pylint: disable=broad-except
+                    logger.debug(traceback.format_exc())
+                    logger.warning(f"Failed to parse DBT column glossary {key}: {exc}")
+
+            return columns_glossary_list or None
+
+        except Exception as exc:  # pylint: disable=broad-except
+            logger.debug(traceback.format_exc())
+            logger.warning(f"Failed to parse DBT column glossary : {exc}")
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/database/dbt/metadata.py:654]: should this yield an `StackTraceError` as well?

File: ingestion/src/metadata/ingestion/source/database/dbt/dbt_service.py [hunk 4]
```diff
@@ -229,6 +240,18 @@ def process_dbt_descriptions(self, data_model_link: DataModelLink):
...
         """
 
+    @abstractmethod
+    def yield_dbt_glossary(self, data_model_link: DbtObjects):
+        """
+        Method to patch DBT glossary
+        """
+
+    @abstractmethod
+    def yield_dbt_tier(self, data_model_link: DbtObjects):
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/database/dbt/dbt_service.py:250]: qq: why do we need a special tier step? Don't we have a yield_tags already there? Same with glossary, in the end it's just a tag with a different source. Should we be able to just reuse the same method but yielding extra details?

File: ingestion/src/metadata/ingestion/source/database/dbt/metadata.py (hunk 3 above)
Comment (SumanMaharana) [reply ingestion/src/metadata/ingestion/source/database/dbt/metadata.py:654]: Here, it's parsing/getting the glossary information passed on the `manifest.json` file from the `dbt` and returns the list of column `glossary terms` to the `yield_data_models` in the source and its processed there;  also its not a `generator` method. so it doesn't yield `StackTraceError`. Please correct me if i'm thinking something wrong here.

File: ingestion/src/metadata/ingestion/sink/metadata_rest.py [hunk 5]
```diff
@@ -518,6 +519,37 @@ def write_life_cycle_data(self, record: OMetaLifeCycleData) -> Either[Entity]:
             )
         )
 
+    @_run_dispatch.register
+    def write_tags_data(self, record: OMetaGlossariesAndTiersData) -> Either[Entity]:
+        """
```
Comment (SumanMaharana) [inline ingestion/src/metadata/ingestion/sink/metadata_rest.py:524]: change to OmetaPatchTags

File: ingestion/src/metadata/ingestion/sink/metadata_rest.py [hunk 6]
```diff
@@ -518,6 +519,37 @@ def write_life_cycle_data(self, record: OMetaLifeCycleData) -> Either[Entity]:
...
+        Ingest the tags data
+        """
+
+        if record.tag_labels:
+
+            # Patch table glossary or tiers
+            self.metadata.patch_tags(
+                entity=record.entity, source=record.table, tag_labels=record.tag_labels
```
Comment (SumanMaharana) [inline ingestion/src/metadata/ingestion/sink/metadata_rest.py:532]: change
record.entity > entity_type
record.table > entity

**Review by OnkarVO7 — CHANGES_REQUESTED** [review]
Apart from the comment let's add the all the tags ( dbt tags, glossaries and tiers) into a single list and process them while ingesting the data model itself.
This way we'll reduce some of the redundant code to patch tags

**Review by OnkarVO7 — APPROVED** [review]

**Review by OnkarVO7 — APPROVED** [review]
